torch/csrc/autograd/autograd.cpp

```diff
@@ -0,0 +1,138 @@
+#include <torch/csrc/autograd/autograd.h>
+#include <torch/csrc/autograd/variable.h>
+
+#include <torch/csrc/autograd/edge.h>
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/autograd/function.h>
+
+namespace torch {
+namespace autograd {
+
+variable_list _make_grads(
+    const variable_list& outputs,
+    const variable_list& grad_outputs) {
+  size_t num_tensors = outputs.size();
+  size_t num_gradients = grad_outputs.size();
+  variable_list new_grads;
+  new_grads.reserve(num_tensors);
+  if (grad_outputs.empty()) {
+    for (const Variable& output : outputs) {
+      if (output.requires_grad()) {
+        TORCH_CHECK(
+            output.numel() == 1,
+            "grad can be implicitly created only for scalar outputs");
+        new_grads.emplace_back(at::ones_like(output));
+      }
+    }
+  } else {
+    TORCH_CHECK(
+        num_tensors == num_gradients,
+        "got %ld tensors and %ld "
+        "gradients",
+        num_tensors,
+        num_gradients);
+    for (size_t i = 0; i < outputs.size(); ++i) {
+      const Variable& output = outputs[i];
+      const Variable& grad_output = grad_outputs[i];
+      if (!grad_output.defined()) {
+        if (output.requires_grad()) {
+          TORCH_CHECK(
+              output.numel() == 1,
+              "grad can be implicitly created only for scalar outputs");
+          new_grads.emplace_back(at::ones_like(output));
+        }
+      } else {
+        // grad output is defined, just append to the new_grads
+        new_grads.emplace_back(grad_output);
+      }
+    }
+  }
+  return new_grads;
+}
+variable_list run_backward(
+    const variable_list& outputs,
+    const variable_list& grad_outputs,
+    bool keep_graph,
+    bool create_graph,
+    const variable_list& inputs) {
+  size_t num_tensors = outputs.size();
+  edge_list roots;
+  roots.reserve(num_tensors);
+  for (size_t i = 0; i < num_tensors; i++) {
+    const Variable& output = outputs[i];
+    auto gradient_edge = output.gradient_edge();
+    TORCH_CHECK(
+        gradient_edge.function,
+        "element %d of tensors does not require grad and does not have a grad_fn",
+        i);
```

Comment:
This is not the correct syntax for `TORCH_CHECK`. `"element ", i, " of tensors" ...`